spec/unleash/constraint_spec.rb

```diff
@@ -72,5 +72,306 @@
+        session_id: 'verylongsesssionid',
+        remote_address: '127.0.0.1',
+        properties: {
+          env: '3.1.41-gamma'
+        }
+      }
+      context = Unleash::Context.new(context_params)
+
+      constraint = Unleash::Constraint.new('env', 'SEMVER_GT', '3.1.41-beta')
+      expect(constraint.matches_context?(context)).to be_truthy
+    end
+
+    it 'matches based on property SEMVER_LT value' do
+      context_params = {
+        user_id: '123',
+        session_id: 'verylongsesssionid',
+        remote_address: '127.0.0.1',
+        properties: {
+          env: '3.1.41-alpha'
+        }
+      }
+      context = Unleash::Context.new(context_params)
+
+      constraint = Unleash::Constraint.new('env', 'SEMVER_LT', '3.1.41-beta')
+      expect(constraint.matches_context?(context)).to be_truthy
+    end
+
+    it 'matches based on property DATE_AFTER value' do
+      context_params = {
+        user_id: '123',
+        session_id: 'verylongsesssionid',
+        remote_address: '127.0.0.1',
+        currentTime: '2022-01-30T13:00:00.000Z'
+      }
+      context = Unleash::Context.new(context_params)
+
+      constraint = Unleash::Constraint.new('currentTime', 'DATE_AFTER', '2022-01-29T13:00:00.000Z')
+      expect(constraint.matches_context?(context)).to be_truthy
+
+      constraint = Unleash::Constraint.new('currentTime', 'DATE_AFTER', '2022-01-31T13:00:00.000Z')
+      expect(constraint.matches_context?(context)).to be_falsey
+    end
+
+    it 'matches based on property DATE_BEFORE value' do
+      context_params = {
+        user_id: '123',
+        session_id: 'verylongsesssionid',
+        remote_address: '127.0.0.1',
+        currentTime: '2022-01-30T13:00:00.000Z'
+      }
+      context = Unleash::Context.new(context_params)
+
+      constraint = Unleash::Constraint.new('currentTime', 'DATE_BEFORE', '2022-01-29T13:00:00.000Z')
+      expect(constraint.matches_context?(context)).to be_falsey
+
+      constraint = Unleash::Constraint.new('currentTime', 'DATE_BEFORE', '2022-01-31T13:00:00.000Z')
+      expect(constraint.matches_context?(context)).to be_truthy
+    end
+
+    it 'matches based on case insensitive property when operator is uppercased' do
+      context_params = {
+        user_id: '123',
+        session_id: 'verylongsesssionid',
+        remote_address: '127.0.0.1',
+        properties: {
+          env: 'development'
+        }
+      }
+      context = Unleash::Context.new(context_params)
+      constraint = Unleash::Constraint.new('env', 'STR_STARTS_WITH', ['DEV'], case_insensitive: true)
+      expect(constraint.matches_context?(context)).to be_truthy
+
+      constraint = Unleash::Constraint.new('env', 'STR_ENDS_WITH', ['MENT'], case_insensitive: true)
+      expect(constraint.matches_context?(context)).to be_truthy
+
+      constraint = Unleash::Constraint.new('env', 'STR_CONTAINS', ['LOP'], case_insensitive: true)
+      expect(constraint.matches_context?(context)).to be_truthy
```

Comment:
Also test when the value is a string, and not just a list of strings.
```suggestion
      constraint = Unleash::Constraint.new('env', 'STR_CONTAINS', ['LOP'], case_insensitive: true)
      expect(constraint.matches_context?(context)).to be_truthy
      
      constraint = Unleash::Constraint.new('env', 'STR_CONTAINS', 'LOP', case_insensitive: true)
      expect(constraint.matches_context?(context)).to be_truthy
```

Reply:
So this is my bad, I don't think the code is reflecting the intent properly here. This operator shouldn't ever receive single values and an operation on them is undefined in the spec. I think I'm going to take your suggestion using the dispatch in your branch and see if I can make this case impossible to get to with some proper validation

Reply:
So the case described here is actually undefined in the spec and the code was too permissive. I've added validation in the constraint constructor to reject invalid datatypes